File: BackEnd/src/main/java/chuchu/runnerway/course/entity/Course.java
package chuchu.runnerway.course.entity;

import chuchu.runnerway.course.dto.request.UserCourseRegistRequestDto;
import chuchu.runnerway.member.domain.Member;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.ColumnDefault;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;

@Entity
@Getter
@Setter
public class Course {

    public Course() {
    }

    @Builder
    public Course(Long courseId, CourseImage courseImage, String name, String address, String content, Long count, int level, int averageSlope, int averageDownhill, LocalDateTime averageTime, double courseLength, CourseType courseType, LocalDateTime registDate, double averageCalorie, double lat, double lng, Member member) {
        this.courseId = courseId;
        this.courseImage = courseImage;
        this.name = name;
        this.address = address;
        this.content = content;
        this.count = count;
        this.level = level;
        this.averageSlope = averageSlope;
        this.averageDownhill = averageDownhill;
        this.averageTime = averageTime;
        this.courseLength = courseLength;
        this.courseType = courseType;
        this.registDate = registDate;
        this.averageCalorie = averageCalorie;
        this.lat = lat;
        this.lng = lng;
        this.member = member;
    }

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "course_id")
    private Long courseId;

    // Member가 생기면 Member 넣어주기
    @ManyToOne
    @JoinColumn(name = "member_id")
    private Member member;

    @Column(name = "name", nullable = false)
    private String name;

    @Column(name = "address", nullable = false)
    private String address;

    @Column(name = "content", length = 500)
    private String content;

    @Column(name = "count")
    @ColumnDefault("0")
    private Long count;

    @Column(name = "level")
    private int level;

    @Column(name = "average_slope")
    private int averageSlope;

    @Column(name = "average_downhill")
    private int averageDownhill;

    @Column(name = "average_time")
    private LocalDateTime averageTime;

    @Column(name = "course_length")
    private double courseLength;

    @Column(name = "courseType", nullable = false)
    @Enumerated(EnumType.STRING)
    private CourseType courseType;

    @Column(name = "regist_date")
    @CreationTimestamp
    private LocalDateTime registDate;

    @Column(name = "average_calorie", nullable = false)
    private double averageCalorie;

    @Column(name = "lat", nullable = false)
    private double lat;

    @Column(name = "lng", nullable = false)
    private double lng;

    @OneToOne(mappedBy = "course", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JsonManagedReference
    private CourseImage courseImage;

    public void updateUserCourse(UserCourseRegistRequestDto userCourseRegistRequestDto, Member member) {
        this.name = userCourseRegistRequestDto.getName();                   // 코스명
        this.address = userCourseRegistRequestDto.getAddress();             // 코스 시작 주소
        this.content = userCourseRegistRequestDto.getContent();             // 코스 한 줄평
        this.member = member;
        this.level = userCourseRegistRequestDto.getLevel();                 // 코스 난이도
        this.averageSlope = userCourseRegistRequestDto.getAverageSlope();   // 평균 경사도
        this.averageDownhill = userCourseRegistRequestDto.getAverageDownhill(); // 평균 내리막도
        this.averageTime = userCourseRegistRequestDto.getAverageTime();     // 예상 평균 소요 시간
        this.courseLength = userCourseRegistRequestDto.getCourseLength();   // 코스 전체 길이
        this.courseType = userCourseRegistRequestDto.getCourseType();       // 코스 타입
        this.averageCalorie = userCourseRegistRequestDto.getAverageCalorie(); // 예상 평균 소모 칼로리
        this.lat = userCourseRegistRequestDto.getLat();                     // 위도
        this.lng = userCourseRegistRequestDto.getLng();                     // 경도
    }
}

File: frontend/lib/models/course.dart
import 'course_image.dart';

class Course {
  final int courseId;
  final String name;
  final String address;
  final String? content;
  final int count;
  final int level;
  final double courseLength;
  final double lat;
  final double lng;
  final double? distance;
  final double? averageSlope;
  final String? averageTime;
  final double? averageCalorie;
  final String? courseType;
  final CourseImage? courseImage;
  final double? averageDownhill;
  final int? memberId;
  final String? memberNickname;
  final DateTime? registDate;
  final int? recordId;

  Course({
    required this.courseId,
    required this.name,
    required this.address,
    this.content,
    required this.level,
    required this.count,
    required this.courseLength,
    required this.lat,
    required this.lng,
    this.courseImage,
    this.distance,
    this.averageSlope,
    this.averageTime,
    this.averageCalorie,
    this.courseType,
    this.averageDownhill,
    this.memberId,
    this.memberNickname,
    this.registDate,
    this.recordId,
  });

  // JSON 데이터를 파싱하여 Course 객체 생성
  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      courseId: json['courseId'],
      name: json['name'],
      address: json['address'],
      content: json['content'],
      level: json['level'],
      count: json['count'],
      courseLength: (json['courseLength'] as num).toDouble(),
      lat: (json['lat'] as num).toDouble(),
      lng: (json['lng'] as num).toDouble(),
      courseImage: json['courseImage'] != null
          ? CourseImage.fromJson(json['courseImage'])
          : null,
      distance: json['distance'] != null
          ? (json['distance'] as num).toDouble()
          : null,
      averageSlope: json['averageSlope'] != null
          ? (json['averageSlope'] as num).toDouble()
          : null,
      averageTime: json['averageTime'],
      averageCalorie: json['averageCalorie'] != null
          ? (json['averageCalorie'] as num).toDouble()
          : null,
      courseType: json['courseType'],
      averageDownhill: json['averageDownhill'] != null
          ? (json['averageDownhill'] as num).toDouble()
          : null,
      memberId: json['memberId'],
      memberNickname: json['memberNickname'],
      registDate: json['registDate'] != null
          ? DateTime.parse(json['registDate'])
          : null,
    );
  }

  // copyWith 메서드
  Course copyWith({
    int? courseId,
    String? name,
    String? address,
    String? content,
    int? count,
    int? level,
    double? courseLength,
    double? lat,
    double? lng,
    double? distance,
    double? averageSlope,
    String? averageTime,
    double? averageCalorie,
    String? courseType,
    CourseImage? courseImage,
    double? averageDownhill,
    int? memberId,
    String? memberNickname,
    DateTime? registDate,
  }) {
    return Course(
      courseId: courseId ?? this.courseId,
      name: name ?? this.name,
      address: address ?? this.address,
      content: content ?? this.content,
      count: count ?? this.count,
      level: level ?? this.level,
      courseLength: courseLength ?? this.courseLength,
      lat: lat ?? this.lat,
      lng: lng ?? this.lng,
      distance: distance ?? this.distance,
      averageSlope: averageSlope ?? this.averageSlope,
      averageTime: averageTime ?? this.averageTime,
      averageCalorie: averageCalorie ?? this.averageCalorie,
      courseType: courseType ?? this.courseType,
      courseImage: courseImage ?? this.courseImage,
      averageDownhill: averageDownhill ?? this.averageDownhill,
      memberId: memberId ?? this.memberId,
      memberNickname: memberNickname ?? this.memberNickname,
      registDate: registDate ?? this.registDate,
    );
  }

  // toJson 메서드 추가
  Map<String, dynamic> toJson() {
    return {
      'courseId': courseId,
      'name': name,
      'address': address,
      'content': content,
      'level': level,
      'count': count,
      'courseLength': courseLength,
      'lat': lat,
      'lng': lng,
      'courseImage': courseImage?.toJson(), // courseImage가 있다면 toJson 사용
      'distance': distance,
      'averageSlope': averageSlope,
      'averageTime': averageTime,
      'averageCalorie': averageCalorie,
      'courseType': courseType,
      'averageDownhill': averageDownhill,
      'memberId': memberId,
      'memberNickname': memberNickname,
      'registDate': registDate?.toIso8601String(), // DateTime을 ISO 8601 형식으로 변환
    };
  }

  // toString 메서드 오버라이드
  @override
  String toString() {
    return 'Course(courseId: $courseId, name: $name, address: $address, '
        'content: $content, count: $count, level: $level, courseLength: $courseLength, '
        'lat: $lat, lng: $lng, distance: $distance, averageSlope: $averageSlope, '
        'averageTime: $averageTime, averageCalorie: $averageCalorie, courseType: $courseType, '
        'averageDownhill: $averageDownhill, memberId: $memberId, memberNickname: $memberNickname, '
        'registDate: $registDate)';
  }
}
